### ISSUE TITLE
IMap.putAll() ignores key locks

### ISSUE THREAD

**Issue by neilstevenson** [hazelcast/hazelcast#10935]
If a key is locked, `putAll()` allows it to be updated. Tried in 3.9-EA.

The documentation  "_It could be that in case of failure some of the key/value-pairs get written, while others are not._" would seem to allow that locked keys should not be updated

**Comment by Donnerbart** [comment]
Reproducer:
```java
public class Main {

    public static void main(String[] args) throws Exception {
        HazelcastInstance hz1 = Hazelcast.newHazelcastInstance();
        HazelcastInstance hz2 = Hazelcast.newHazelcastInstance();

        IMap<Integer, Integer> map1 = hz1.getMap("myMap");
        IMap<Integer, Integer> map2 = hz2.getMap("myMap");

        map1.put(1, 0);
        map1.lock(1);

        Map<Integer, Integer> map = new HashMap<>();
        map.put(1, 23);
        map.put(2, 42);
        map2.putAll(map);

        map1.unlock(1);

        System.out.println("1: " + map1.get(1) + ", 2: " + map1.get(2));
        System.out.println("1: " + map2.get(1) + ", 2: " + map2.get(2));

        Hazelcast.shutdownAll();
    }
}
```
Output:
```
1: 23, 2: 42
1: 23, 2: 42
```

**Comment by gurbuzali** [comment]
Currently if an entry is locked and you call `put` on it, the call blocks. What should be the behavior in case of `putAll` ? 
`putAll` splits the entries to partitions, in case of a locked entry only one operation will be blocked and the rest will be put, is this the expected behavior ?
Another approach may be to exclude locked keys and put the rest.

**Comment by tombujok** [comment]
The same problem exists with EntryProcessor, which ignores locks if run for multiple keys.
I think we should address both problems in 4.0 where we can make breaking changes.

**Comment by neilstevenson** [comment]
Impacts on Jet, https://github.com/hazelcast/hazelcast-jet/blob/master/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/connector/HazelcastWriters.java#L71

**Comment by neilstevenson** [comment]
Looks like `IMap.clear()` may also ignore locks.

**Comment by enesakar** [comment]
putAll has performance advantage over puts in a loop. 
I am worried fixing this may kill the performance advantage.

**Comment by pveentjer** [comment]
This seems like a bug and doesn't need to be bound to the 4.0 release.

**Comment by mmedenjak** [comment]
I agree this is a bug but as it changes behavior, it might not be expected by some users.
